### PR TITLE
[BugFix] Do not let an apportioned row count decide whether a cross-published rowset exists

### PR DESCRIPTION
**Pull request by luohaha**
## Why I'm doing:

Part of #77653. Cross publish hands a SPLIT child the parent's whole payload and then leaves each
side to work out which rows are actually its own. Every place that gets that wrong is the same bug
wearing a different hat; here the wrong answer is given by a statistic, and a row disappears.

A row written while a tablet-split job is in flight is **silently discarded**. The `INSERT` returns
success and the label reaches `VISIBLE`; after the split finishes an unpredicated `SELECT COUNT(*)`
is short by exactly that row. No fault injection, no concurrency, no node involvement. Reproduced
10/10 on `main`.

### What actually happens

A write that has already put its data on the old tablet but has not committed yet holds **no
partition version**, so the split's only barrier — `runPreparingJob`'s `commitVersion ==
visibleVersion + 1` wait — cannot see it. It then commits at a version *above* the split's reserved
version. Millisecond timeline from one loss (partition 150561):

```
15.027  probe txn 1183922 PREPARE  -> starts writing to OLD tablet 150562
15.027  split job 150577 created, PENDING
15.034  job -> PREPARING: runPendingJob's updateNextVersions() takes version 3
15.061  probe txn 1183922 COMMIT   -> gets partition version 4
15.035  split publishes base_version=2 -> new_version=3
16.097  probe txn 1183922 VISIBLE at version 4 (waited 1013ms for v3), returns success
```

FE handles this correctly: `Utils.processTablets` finds the old tablet in
`TabletReshardJobMgr`'s resharding registry and redirects the publish to the K new tablets as a
`SPLITTING` cross publish. On a dedicated cluster (idle `tablet_reshard_cross_publish_splitting_total`
delta = 0) every round performs **exactly 4** conversions, one per new tablet — so the redirect and
the BE-side conversion both run.

The row dies inside the conversion. `convert_txn_log_for_splitting` range-restricts the shared rowset
per sibling (`update_rowset_ranges`) and then apportions its stats:

```cpp
num_rows / split_count + (split_index < num_rows % split_count ? 1 : 0)
```

This keys off the **split index** and knows nothing about which sub-range the rows occupy. For a
1-row write split 4 ways, index 0 is handed 1 row and indexes 1..3 are handed **0** — and a 0-row
rowset is dropped outright by both txn log appliers:

* `NonPrimaryKeyTxnLogApplier::apply_write_log` (`txn_log_applier.cpp:1123`) only attaches the rowset
  when `num_rows() > 0 || has_delete_predicate()`
* `PrimaryKeyTxnLogApplier::apply_write_log` (`txn_log_applier.cpp:582`) returns early on the same
  condition, so the rowset is never attached and the PK index is never updated

So the row survives only if its sub-range happens to be one that got a non-zero share. Confirmed by
three discriminating runs — the outcome tracks the apportioned value, not the index:

| probe key falls in | that index apportioned | result |
|---|---|---|
| last sub-range (index 3) | 0 | **lost** |
| first sub-range (index 0) | 1 | survives |
| 4 probes, one per sub-range (4 % 4 == 0, everyone gets 1) | 1 | all 4 survive |

Index 3 is fine when it receives 1 and loses the row when it receives 0, which rules out anything
index-specific.

**Scope.** The split's *own* path is not affected: it distributes per-rowset stats through
`range_source_stats`, which is range aware. Verified separately — a 1-row rowset committed and
visible *before* a fully serialized split survives. Only the cross-publish path uses the pure
index-based apportionment.

## What I'm doing:

**Stop a statistic from deciding whether data exists.** Both single-log appliers skipped an
`op_write` whose `rowset.num_rows()` was 0. That count is apportioned per sibling on a cross publish,
so a rowset whose segments hold this tablet's rows can legitimately arrive at 0 — and then its
segments are dropped. Presence now keys off what the op carries (segments, del files, del sstables, a
delete predicate). The batch applier already worked this way
(`NonPrimaryKeyBatchZeroNumRowsKeepsSegmentAndUid`); this brings the single-log path into line.

This is not a behaviour change for an ordinary publish: a writer that produced no rows produces no
segments either, so the two conditions agree everywhere except after the cross-publish apportionment.

**Make the apportionment range aware.** Classify each rowset's key envelope (min/max over its
segments' `sort_key_min`/`sort_key_max`) against the sibling's range:

| classification | meaning | apportionment |
|---|---|---|
| `kNo` | envelope lies entirely outside the range — the sibling **provably** owns none of the rows | a true `0`, so the siblings' stats sum to the source instead of spreading phantom rows over everyone |
| `kYes` | envelope intersects the range — the sibling **may** own rows | the plain share |
| `kUnknown` | no sort-key bounds, or bounds whose arity does not match the range's | the plain share, i.e. pre-existing behavior |

`kUnknown` exists because comparing tuples of different arity can flip the boundary case
(`VariantTuple::compare` orders a shorter prefix-equal tuple *below* its longer form, so envelope
`(100)` vs lower bound `(100, MIN)` would read as "range entirely above the envelope" even though key
100 is the range's first key) — a mismatch must never become a false `kNo`.

**Conservation.** The shares sum to the source exactly. An earlier revision of this PR rounded a
`kYes` share up to 1 so the rowset would survive the `num_rows == 0` skip; with presence decoupled
that floor is gone, and with it the over-count of up to `split_count - 1` rows it caused.

I deliberately did **not** compute the true per-sibling row count, which would mean reading the
rowset's segment index on the publish hot path. `kNo` is a sound "owns nothing" proof because the
envelope is a superset of the rowset's real keys; `kYes` remains "may own rows", not "does".

Fixes StarRocks/StarRocksTest#11729

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [x] Miscellaneous: a cross-published rowset's per-sibling `num_rows`/`data_size` are now derived from
      range overlap. A sibling that owns nothing reports 0 (previously a non-zero share), and a
      sibling that may own rows never reports 0. Totals stay exact except when a multi-row envelope
      spans several siblings and the row count is below the split count, where they can exceed the
      source by at most `split_count - 1`.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
  - [ ] This pr needs auto generate documentation
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 4.1
  - [ ] 4.0
  - [ ] 3.5

Range distribution / tablet reshard is main-only, so no backport is needed.

## Tests

New cases in `be/test/storage/lake/tablet_reshard_helper_test.cpp`:

- `test_classify_rowset_range_overlap` — single-key rowset lands in exactly one of a set of disjoint
  ranges; inclusive-lower / exclusive-upper boundary cases; a straddling span overlaps; a multi-segment
  envelope spans its segments; `(-inf, +inf)` contains everything; and every undecidable input
  (no segments, missing bounds, one bound-less segment among several) degrades to `kUnknown`.
- `test_update_rowset_data_stats_never_zeroes_an_overlapping_rowset` — a 1-row rowset split 4 ways
  gives every `kYes` sibling at least 1 row and 1 byte (the regression).
- `test_update_rowset_data_stats_zeroes_a_non_overlapping_rowset` — a `kNo` sibling reports 0.
- `test_update_rowset_data_stats_unknown_overlap_keeps_legacy` — `kUnknown` reproduces the old
  `[1, 0, 0, 0]` apportionment exactly.
- `test_update_rowset_data_stats_conserves_when_rows_exceed_split_count` — with `num_rows >=
  split_count` the apportionment is untouched and totals conserve exactly.

The pre-existing `LakeTabletReshardTest.test_split_cross_publish_multi_stmt_batch_keeps_scaled_zero_segment`
stays green untouched: its synthetic segment metas carry no sort-key bounds, so it classifies
`kUnknown` and keeps the legacy per-statement apportionment (and its exact conservation assertions).

E2E: `test_range_reshard_write_loss.py` (PK / DUP) from StarRocks/StarRocksTest#11709.

### CODE REVIEW

**Comment by CelerData-Reviewer** [comment]
@codex review

File: be/src/storage/lake/tablet_reshard_helper.cpp [hunk 1]
```diff
@@ -620,13 +620,32 @@ Status update_rowset_ranges(TxnLogPB* txn_log, const TabletRangePB& range) {
...
+    // the rows of a cross-published write land nowhere if they happen to belong to one of those
+    // siblings, while the transaction still reports VISIBLE.
+    //
+    // Round a non-empty rowset up to 1 instead of conserving the total exactly: over-counting a
+    // stat can only skew compaction scoring and SHOW-level estimates, whereas under-counting to
+    // zero destroys data. data_size gets the same treatment -- a zero data_size makes
+    // Rowset::data_size_after_deletion() return 0 and feeds the same scoring.
+    auto apportion = [split_count, split_index](int64_t value) {
+        const int64_t scaled = value / split_count + (split_index < value % split_count ? 1 : 0);
+        return value > 0 ? std::max<int64_t>(1, scaled) : scaled;
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/tablet_reshard_helper.cpp]: Thanks — the structural reading is right (`RowsetUpdateState::rewrite_segment`'s guard at `rowset_update_state.cpp:494` keys off `rowset_meta.num_rows()`, which this clamp raises from 0 to 1), but I don't think the conclusion that this PR *introduces* the exposure holds. That path is already taken today.

The guard fires only for a row-mode partial update (`has_txn_meta() && rewrite_segments_meta_size() > 0`), and `rewrite_segments_meta` comes from the load's txn log, which `convert_txn_log_for_splitting` copies verbatim to every sibling — so it is non-zero for all K siblings regardless of this change. The only per-sibling variable is `num_rows`:

```
num_rows=8, K=4  ->  [2, 2, 2, 2]   before and after this PR
num_rows=1, K=4  ->  [1, 0, 0, 0]   before
num_rows=1, K=4  ->  [1, 1, 1, 1]   after
```

So a row-mode partial update whose rows all land in one sub-range **already** puts the other three range-empty siblings through `rewrite_segment` whenever `num_rows >= split_count`, which is the common case. This PR makes the `num_rows < split_count` case behave like the case that is already the norm; it does not create a new code path.

Narrowing the clamp to skip partial updates is not a good trade either: it would leave exactly the reported data loss in place for partial-update writes (a 1-row partial update cross-published 4 ways loses its row whenever that row belongs to a zero-apportioned sibling — the same bug this PR fixes).

That said, I have not exercised the partial-update path empirically, so I am not going to claim it is fine on reasoning alone. A TSP build of this PR is compiling; I will run a row-mode partial update inside the loss window against both the unpatched and patched binaries and report the result here. If range-empty partial-update siblings turn out to be genuinely broken, that is a pre-existing bug worth its own fix rather than a reason to keep under-counting `num_rows` to zero — but I'll let the measurement decide.

[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/tablet_reshard_helper.cpp]: Reworked in 165354ff7d7 — and the new shape resolves this finding outright rather than arguing about it.

Two things pushed me off the blanket clamp. First, an existing test (`LakeTabletReshardTest.test_split_cross_publish_multi_stmt_batch_keeps_scaled_zero_segment`) went red on it, and reading that test showed the **multi-statement batch path already handles a zero-scaled statement safely**: the statements are combined into one rowset, so the zero-scaled statement's segment survives and the combined `num_rows` is non-zero. The data loss I measured is the **single-statement** case, where there is no sibling statement to keep the rowset alive. So the clamp perturbed a path that was never broken (and broke its exact stat conservation) while fixing one that was. Second, your point stands that fabricating a row count to force a sibling through the publish path is the wrong lever.

The fix now classifies the rowset's key envelope (min/max over its segments' `sort_key_min`/`sort_key_max`) against the sibling's range:

- **`kNo`** — envelope entirely outside the range ⇒ the sibling provably owns nothing ⇒ a **true 0**, and the appliers drop it. Because the envelope is a superset of the real keys, this is sound.
- **`kYes`** — envelope intersects ⇒ the sibling may own rows ⇒ apportion as before but never below 1.
- **`kUnknown`** — no sort-key bounds, or an arity that does not match the range's ⇒ legacy apportionment byte for byte.

For your scenario specifically: a row-mode partial update whose rows all sit in one sub-range now classifies **`kNo`** on the other siblings, so they get `num_rows == 0`, the appliers skip them, and `rewrite_segment` is never reached. That is strictly *better* than the status quo — today those range-empty siblings receive a non-zero share (whenever `num_rows >= split_count`) and **do** go through `rewrite_segment`. So this PR now reduces that exposure instead of widening it.

Conservation is exact for the losing case (a small rowset has a single-key envelope ⇒ exactly one `kYes` sibling takes all the rows) and whenever `num_rows >= split_count`; it can over-count by at most `split_count - 1` only when a multi-row envelope spans several siblings and the row count is below the split count.

The previously-failing test stays green untouched — its synthetic segment metas carry no sort-key bounds, so it classifies `kUnknown` and keeps the legacy per-statement apportionment and its conservation assertions. New unit tests cover the classifier (boundary inclusivity, straddling spans, multi-segment envelopes, and every undecidable input degrading to `kUnknown`) plus the three apportionment branches.

**Comment by CelerData-Reviewer** [comment]
@codex review

File: be/src/storage/lake/tablet_reshard_helper.cpp [hunk 2]
```diff
@@ -617,16 +617,91 @@ Status update_rowset_ranges(TxnLogPB* txn_log, const TabletRangePB& range) {
...
+
+    // kYes: this sibling may own rows, so never hand it an empty rowset -- the appliers would drop
+    // it and those rows would vanish while the transaction still reported success. Rounding up to 1
+    // can over-count by at most (split_count - 1) rows per rowset, and only for a rowset with fewer
+    // rows than the split count; an over-counted stat merely skews compaction scoring, whereas
+    // under-counting to zero destroys data. kUnknown keeps the pre-existing behavior exactly.
+    const bool never_empty = (overlap == RangeOverlap::kYes);
+    auto apportion = [split_count, split_index, never_empty](int64_t value) {
+        const int64_t scaled = value / split_count + (split_index < value % split_count ? 1 : 0);
+        return (never_empty && value > 0) ? std::max<int64_t>(1, scaled) : scaled;
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/tablet_reshard_helper.cpp]: Measured it, as promised. **The failure this points at is pre-existing and byte-for-byte unaffected by this PR.**

Test: PK range table, 10k rows, split into 4 (boundaries 3073 / 6145 / 9217). Immediately after `ALTER TABLE ... SPLIT TABLET`, a row-mode partial-update stream load updating **two keys in distant sub-ranges** — 100 (sub-range 0) and 9500 (sub-range 3) — so the min/max envelope spans all four siblings and sub-ranges 1 and 2 classify `kYes` while holding none of the updated keys. Exactly your scenario. Same cluster, same data, only the BE binary swapped:

| build | result (3 rounds each) |
|---|---|
| **with this PR** (range-aware apportionment) | `Publish Timeout`, reshard job stuck at `CLEANING`, `v(100)=0 v(9500)=0`, `rows=10000`, `pad_len=64` |
| **without this PR** (plain main BE) | `Publish Timeout`, reshard job stuck at `CLEANING`, `v(100)=0 v(9500)=0`, `rows=10000`, `pad_len=64` |

Identical, 3/3 both sides. So the clamp is not what breaks it — consistent with the arithmetic: 2 rows split 4 ways already apportions `[1, 1, 0, 0]` **without** this change, so sibling 1 is row-empty with `num_rows == 1` on main today.

Two details that refine the diagnosis:

- The observable failure is a **publish timeout with the update lost**, not a corrupt segment: `rows` stays 10000 and the unmodified `pad` column is intact (64 chars) on both builds. So the "rewrite the original segment with empty replacement columns → inconsistent segment" mechanism does not appear to materialise; something in the publish stalls instead.
- Contrast with the **single-key** partial update in the same window, which this PR does fix: without it → `Publish Timeout` + update lost; with it → `Success` + `v=777` applied. So this PR strictly improves the partial-update case it covers and leaves the multi-sub-range case exactly as it was.

I've therefore left the design as is and pinned the limitation explicitly in `test_classify_rowset_range_overlap_sparse_envelope_is_conservative`, with a comment recording that `kNo` must be a proof and that deciding a sparse envelope exactly would mean reading the rowset's segment index on the publish hot path.

The pre-existing multi-sub-range failure is a real bug and I'll report it separately rather than widen this PR.

**Comment by CelerData-Reviewer** [comment]
@codex review

File: be/src/storage/lake/tablet_reshard_helper.cpp [hunk 3]
```diff
@@ -623,16 +623,87 @@ Status update_rowset_ranges(TxnLogPB* txn_log, const TabletRangePB& range) {
...
+    // the appliers decide a rowset's presence from its segments, so a zero here costs accuracy in
+    // compaction scoring and `SHOW`, never data.
+    auto apportion = [split_count, split_index](int64_t value) {
+        return value / split_count + (split_index < value % split_count ? 1 : 0);
+    };
+
     if (rowset->has_num_rows()) {
-        int64_t num_rows = rowset->num_rows();
-        rowset->set_num_rows(num_rows / split_count + (split_index < num_rows % split_count ? 1 : 0));
+        rowset->set_num_rows(apportion(rowset->num_rows()));
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/tablet_reshard_helper.cpp:717]: Agreed, and this is already fixed — you reviewed `3703dd4`; the fix landed one commit later in `a281ca2`, which keys `RowsetUpdateState::rewrite_segment` off the same payload predicate as the appliers:

```cpp
// be/src/storage/lake/rowset_update_state.cpp
if (!params.op_write.has_txn_meta() || params.op_write.rewrite_segments_meta_size() == 0 ||
    rowset_meta.segment_metas_size() == 0) {
    return Status::OK();
}
```

The scenario you describe — a row-mode partial update whose owning sibling draws a zero share — no longer short-circuits, so the partial segment gets its unmodified columns rewritten in before it is attached. Outside a cross publish the two predicates agree: a writer that produced no rows produces no segments either.

Separately, the zero itself is now avoided at the source: `075a789` floors a segment-carrying rowset's share at 1 (see the reply on the conservation finding below).

File: be/src/storage/lake/txn_log_applier.cpp [hunk 4]
```diff
@@ -1120,7 +1126,10 @@ class NonPrimaryKeyTxnLogApplier : public TxnLogApplier {
     Status apply_write_log(const TxnLogPB_OpWrite& op_write, int64_t txn_id) {
         TEST_ERROR_POINT("NonPrimaryKeyTxnLogApplier::apply_write_log");
         RETURN_IF_ERROR(update_metadata_schema(op_write, txn_id, _metadata, _tablet.tablet_mgr()));
-        if (op_write.has_rowset() && (op_write.rowset().num_rows() > 0 || op_write.rowset().has_delete_predicate())) {
+        // Carries data if it carries segments -- see the note on the primary-key applier: num_rows is
+        // apportioned per sibling on a cross publish and must not decide whether the rowset exists.
+        if (op_write.has_rowset() &&
+            (op_write.rowset().segment_metas_size() > 0 || op_write.rowset().has_delete_predicate())) {
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/txn_log_applier.cpp]: Already fixed — you reviewed `3703dd4`; `a281ca2` moved `build_rssid_remap` onto the identical payload predicate one commit later:

```cpp
// The predicates below must stay identical to the ones the two appliers use to decide whether an
// op_write carries anything -- they model the same target_id advance.
if (is_pk) {
    included = op_write.rowset().segment_metas_size() > 0 || op_write.dels_meta_size() > 0 ||
               op_write.del_ssts_size() > 0 || op_write.rowset().has_delete_predicate();
} else {
    included = op_write.rowset().segment_metas_size() > 0 || op_write.rowset().has_delete_predicate();
}
```

So the remap and both appliers now advance `target_id` in lockstep and DCG entries stay aligned.

File: be/src/storage/lake/tablet_reshard_helper.cpp [hunk 5]
```diff
@@ -623,16 +623,87 @@ Status update_rowset_ranges(TxnLogPB* txn_log, const TabletRangePB& range) {
...
+                              RangeOverlap overlap) {
     if (split_count <= 1) return;
 
+    if (overlap == RangeOverlap::kNo) {
+        // Proven to own none of the rowset's keys: a true zero, not an apportionment artifact. The
+        // appliers drop the rowset for this sibling, which is exactly right -- and it keeps the
+        // siblings' stats summing to the source instead of spreading phantom rows over everyone.
+        if (rowset->has_num_rows()) rowset->set_num_rows(0);
+        if (rowset->has_data_size()) rowset->set_data_size(0);
+        if (rowset->has_num_dels()) rowset->set_num_dels(0);
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/tablet_reshard_helper.cpp:705]: Real, and `075a789` fixes the part that is reachable and harmful.

Your scenario is exactly right: with the `kNo` siblings zeroed outright, the remainder that plain apportionment hands to the leading indexes can land on a sibling that no longer exists, so a one-row rowset owned by child 3 of 4 leaves *every* sibling at `num_rows == 0` — including the one still carrying the segments. A rowset that carries segments and reports zero rows reads as empty to every consumer of the statistic, so that share is now floored at 1:

```cpp
const bool carries_segments = rowset->segment_metas_size() > 0;
...
const int64_t scaled = apportion(rowset->num_rows());
rowset->set_num_rows((carries_segments && rowset->num_rows() > 0) ? std::max<int64_t>(1, scaled) : scaled);
```

The floor is keyed off the payload rather than the classifier verdict, so a `kNo` sibling's zero still stands (it genuinely owns nothing, which is *more* accurate than the pre-PR behaviour of handing it the leading index's whole share) and a stats-only rowset with no segments keeps the pre-existing apportionment byte for byte.

On full conservation across siblings: I don't think it can be restored here. `update_rowset_data_stats` runs inside each sibling's own publish and sees only its own range and `split_index` — it cannot enumerate which of its peers the classifier ruled out, so there is no local redistribution to compute. Exact per-sibling counters need the actual row selection during cross publish, which is CP-3/CP-4 of #77653; until then these remain estimates in both directions (the floor over-counts by at most `split_count - 1`, and only for a rowset holding fewer rows than the split count). No correctness consumer depends on them — presence is decided by the segments now, and the PK compaction policy only accumulates them into scores (`primary_key_compaction_policy.cpp:204,364,442`), with `num_dels` still clamped to `num_rows`.

**Comment by CelerData-Reviewer** [comment]
@codex review

File: be/src/storage/lake/tablet_reshard_helper.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/tablet_reshard_helper.cpp:705]: Correction to my previous reply: I reverted the floor. It was the wrong fix.

`LakeTabletReshardTest.test_split_cross_publish_multi_stmt_batch_keeps_scaled_zero_segment` already pins conservation as the contract:

```cpp
// tablet_reshard_test.cpp:3439
EXPECT_EQ(10, child0_rowset.num_rows() + child1_rowset.num_rows());
EXPECT_EQ(110, child0_rowset.data_size() + child1_rowset.data_size());
```

A one-row statement scaling to 0 on `split_index=1` is exactly the case that test walks, and the floor turned the sum into 11. So the plain apportionment stays and the `kYes`/`kUnknown` shares are exact again.

That leaves your finding standing, narrowed to the `kNo` branch: zeroing a sibling proven to own nothing makes the sibling total a lower bound rather than an exact split. I still don't think it can be redistributed here — `update_rowset_data_stats` runs inside each sibling's own publish and sees only its own range and `split_index`, so it cannot enumerate which peers the classifier ruled out. Exact per-sibling counters need the real row selection during cross publish (CP-3/CP-4 of #77653). Nothing depends on them for correctness: presence is decided by the payload now, and the PK compaction policy only accumulates them into scores.

What did need fixing was the presence predicate itself, and the segment count turned out to be too coarse for it — an empty write also produces a segment, just one holding no rows, so `LakePrimaryKeyCompactionTest.test3` saw that rowset attached and compaction picked up three inputs instead of two. Only the rowset-level `num_rows` is apportioned across siblings; the per-segment counts ride through untouched, so those are what tell the two cases apart:

```cpp
bool rowset_holds_rows(const RowsetMetadataPB& rowset) {
    if (rowset.num_rows() > 0) return true;
    if (rowset.segment_metas_size() == 0) return false;
    bool any_segment_counted = false;
    for (const auto& segment_meta : rowset.segment_metas()) {
        if (segment_meta.num_rows() > 0) return true;
        any_segment_counted |= segment_meta.has_num_rows();
    }
    return !any_segment_counted;  // legacy back-fill carries no count; nothing proves it empty
}
```

BE UT is green on this (3426 tests).

File: be/src/storage/lake/tablet_reshard_helper.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/tablet_reshard_helper.cpp:717]: Follow-up: the gate is still keyed off the payload, but its final form is additive rather than a replacement, since the segment count alone turned out to be too coarse elsewhere in the change:

```cpp
// be/src/storage/lake/rowset_update_state.cpp
if (!params.op_write.has_txn_meta() || params.op_write.rewrite_segments_meta_size() == 0 ||
    (rowset_meta.num_rows() == 0 && rowset_meta.segment_metas_size() == 0)) {
    return Status::OK();
}
```

The case you raised is unchanged: a row-mode partial update whose owning sibling drew a zero share still reaches the rewrite, so the partial segment gets its unmodified columns rewritten in before it is attached.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by luohaha** [comment]
CI status note: `ADMIT TEST` is red here for a reason unrelated to this PR.

It dies mid-run on `test_low_cardinality_join` with a burst of `pymysql.err.InterfaceError: (0, '')` across every worker at one instant, and the harness then records `Ran 0 tests in 1230.034s` — the shared `ci-admit-cloud` cluster goes away, so no case results are collected and there is no `[FAIL]` list anywhere in the log.

The same thing happens on an unrelated PR against the same shared cluster:

| | this PR (job 95140222070) | #77562 `fix/iceberg-rest-419-token-expiry` (job 95125795609) |
|---|---|---|
| cluster | `ci-admit-cloud` | `ci-admit-cloud` |
| died during | `test_low_cardinality_join` | `test_low_cardinality_join` |
| cases finished | 368 | 370 |
| harness result | `Ran 0 tests in 1230.034s` | `Ran 0 tests in 1229.980s` |

Everything else is green on this head: `BUILD`, `BE UT` (3426 tests), `SQL-Tester`, `DCO`, `BE Architecture`, `Clang-Format`, `Clang-Tidy`, `Schema Compatibility`, coverage.

**Review by kevincai — APPROVED** [review]

**Review by xiangguangyxg — APPROVED** [review]

**Comment by luohaha** [comment]
https://github.com/Mergifyio backport branch-4.1
